File: Sources/windows/onscreen/features/gets_angry_when_meeting_other_cats.py
from datetime import datetime, timedelta
from typing import List
from yage.models.animations import EntityAnimation
from yage.models.capability import Capability
from yage.models.collisions import Collision
from yage.models.entity_state import EntityState
from yage.utils.logger import Logger


class GetsAngryWhenMeetingOtherCats(Capability):
    def __init__(self, subject):
        super().__init__(subject)
        self.unlock_date = None

    def update(self, collisions, time):
        if self.unlock_date and self.unlock_date > datetime.now():
            return
        self.unlock_date = None
        self.do_update(collisions, time)

    def do_update(self, collisions, time):
        if self.subject.state != EntityState.MOVE:
            return
        if not self.is_touching_another_cat(according_to=collisions):
            return
        self.get_angry()

    def get_angry(self):
        Logger.log(self.tag, "Getting angry!")
        self.subject.animations_scheduler.load_animation(
            EntityAnimation("angry"),
            4
        )
        self.unlock_date = datetime.now() + timedelta(seconds=30)

    def is_touching_another_cat(self, according_to: List[Collision]) -> bool:
        for collision in according_to:
            if collision.other.species.id.startswith("cat"):
                return True
        return False

File: Sources/macOS/Sources/OnScreen/Features/GetsAngryWhenMeetingOtherCats.swift
import Foundation
import Schwifty
import Yage

class GetsAngryWhenMeetingOtherCats: Capability {
    override func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard subject?.state == .move else { return }
        guard isTouchingAnotherCat(accordingTo: collisions) else { return }
        getAngry()
    }
    
    private func getAngry() {
        guard let subject else { return }
        Logger.log(subject.id, "Getting angry!")
        subject.set(state: .action(action: .angry, loops: 4))
        isEnabled = false
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
            self?.stopBeingAngry()
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 30) { [weak self] in
            self?.isEnabled = true
        }
    }
    
    private func stopBeingAngry() {
        guard let subject = subject as? PetEntity else { return }
        guard case .action(let action, _) = subject.state else { return }
        guard action.id == "angry" else { return }
        subject.direction = .init(dx: -1, dy: 0)
        subject.resetSpeed()
        subject.set(state: .move)
    }

    private func isTouchingAnotherCat(accordingTo collisions: Collisions) -> Bool {
        collisions.contains {
            $0.other?.species.id.hasPrefix("cat") == true
        }
    }
}

private extension EntityAnimation {
    static let angry = EntityAnimation(id: "angry")
}
